File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/G-Кассы/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N + M)
    Memory Complexity O(M)
    """
    n: int = int(sys.stdin.readline().rstrip())
    seconds: list[int] = [0] * (24 * 60 * 60)
    for _ in range(n):
        s_h, s_m, s_s, e_h, e_m, e_s = map(int, sys.stdin.readline().rstrip().split())
        start: int = 3600 * s_h + 60 * s_m + s_s
        end: int = 3600 * e_h + 60 * e_m + e_s
        if start == end: seconds[0] += 1
        elif start < end:
            seconds[start] += 1
            seconds[end] -= 1
        else:
            seconds[0] += 1
            seconds[end] -= 1
            seconds[start] += 1
    output: int = 0 
    workers: int = 0
    for second in range(0, len(seconds)):
        workers += seconds[second]
        if workers == n: output += 1
    sys.stdout.write('{}\n'.format(output))


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/G-Кассы/solution.cpp
#include <iostream>
#include <vector>


void solution() {
    /*
    Time Complexity O(N + M)
    Memory Complexity O(M)
    */
    int n;
    std::scanf("%d", &n);
    std::vector<int> seconds (24 * 60 * 60, 0);
    for (int i = 0; i < n; ++i) {
        int sH, sM, sS, eH, eM, eS;
        std::scanf("%d %d %d %d %d %d", &sH, &sM, &sS, &eH, &eM, &eS);
        int start = 3600 * sH + 60 * sM + sS;
        int end = 3600 * eH + 60 * eM + eS;
        if (start == end) ++seconds[0];
        else if (start < end) {
            ++seconds[start];
            --seconds[end];
        } else {
            ++seconds[0];
            --seconds[end]; 
            ++seconds[start];
        }
    }
    int output = 0;
    int workers = 0;
    for (int second = 0; second < seconds.size(); ++second) {
        workers += seconds[second];
        if (workers == n) ++output;
    }
    std::printf("%d\n", output);
}


int main() {
    solution();
}